Signal/src/util/OWSScreenLockUI.m

```diff
@@ -0,0 +1,253 @@
+    if (self.appIsInactive) {
+        // Don't show 'Screen Lock' if app is inactive.
+    } else if (![TSAccountManager isRegistered]) {
+        // Don't show 'Screen Lock' if user is not registered.
+    } else if (!OWSScreenLock.sharedManager.isScreenLockEnabled) {
+        // Don't show 'Screen Lock' if 'Screen Lock' isn't enabled.
+    } else if (self.hasUnlockedScreenLock) {
+        // Don't show 'Screen Lock' if 'Screen Lock' has been unlocked.
+    } else if (!self.appBecameInactiveDate) {
+        // Show 'Screen Lock' if app hasn't become inactive yet (just launched).
+        shouldHaveScreenLock = YES;
+    } else {
+        OWSAssert(self.appBecameInactiveDate);
+
+        NSTimeInterval screenLockInterval = fabs([self.appBecameInactiveDate timeIntervalSinceNow]);
+        NSTimeInterval screenLockTimeout = OWSScreenLock.sharedManager.screenLockTimeout;
+        OWSAssert(screenLockInterval >= 0);
+        OWSAssert(screenLockTimeout >= 0);
+        if (self.appBecameInactiveDate && screenLockInterval < screenLockTimeout) {
+            // Don't show 'Screen Lock' if 'Screen Lock' timeout hasn't elapsed.
+            shouldHaveScreenLock = NO;
+        } else {
+            // Otherwise, show 'Screen Lock'.
+            shouldHaveScreenLock = YES;
+        }
+    }
+
+    // Show 'Screen Protection' if:
+    //
+    // * App is inactive and...
+    // * Either 'Screen Protection' or 'Screen Lock' is enabled.
+    BOOL shouldHaveScreenProtection = (self.appIsInactive
+        && (Environment.preferences.screenSecurityIsEnabled || OWSScreenLock.sharedManager.isScreenLockEnabled));
+
+    BOOL shouldShowBlockWindow = shouldHaveScreenProtection || shouldHaveScreenLock;
+    DDLogVerbose(@"%@, shouldHaveScreenProtection: %d, shouldHaveScreenLock: %d, shouldShowBlockWindow: %d",
+        self.logTag,
+        shouldHaveScreenProtection,
+        shouldHaveScreenLock,
+        shouldShowBlockWindow);
+    self.screenBlockingWindow.hidden = !shouldShowBlockWindow;
+
+    if (shouldHaveScreenLock) {
+        if (!self.isShowingScreenLockUI) {
+            self.isShowingScreenLockUI = YES;
+
+            [OWSScreenLock.sharedManager tryToUnlockScreenLockWithSuccess:^{
+                DDLogInfo(@"%@ unlock screen lock succeeded.", self.logTag);
+                self.isShowingScreenLockUI = NO;
+                self.hasUnlockedScreenLock = YES;
+                [self ensureScreenProtection];
+            }
+                failure:^(NSError *error) {
+                    DDLogInfo(@"%@ unlock screen lock failed.", self.logTag);
+                    self.isShowingScreenLockUI = NO;
+
+                    [self showScreenLockFailureAlertWithMessage:error.localizedDescription];
+                }
+                cancel:^{
+                    DDLogInfo(@"%@ unlock screen lock cancelled.", self.logTag);
+                    self.isShowingScreenLockUI = NO;
+
+                    [self showScreenLockFailureAlertWithMessage:
```

Comment:
Presumably if the user hit cancel, we don't need to show an alert. Can we just go back to the lock screen?

Reply:
Done.